### PR TITLE
Support resolving keyed services from DI in RDF and RDG

### PR DESCRIPTION
**Pull request by captainsafia**
## Description

This PR adds support for generating the appropriate code gen, for both run-time and compile-time generation strategies, to resolve keyed services from dependency injection in minimal APIs.

Fixes https://github.com/dotnet/aspnetcore/issues/49633

## Customer Impact

Support for keyed services was added to the runtime in .NET 8 Preview 7. This change allows customers to leverage keyed services in their minimal APIs and contributes to a more complete user experience.

## Regression?

- [ ] Yes
- [X] No

## Risk

- [ ] High
- [ ] Medium
- [X] Low

Low risk, because the change is additive and doesn't affect pre-existing codepaths around parameter binding

## Verification

- [X] Manual (required)
- [X] Automated

## Packaging changes reviewed?

- [ ] Yes
- [ ] No
- [X] N/A

### CODE REVIEW

File: src/Shared/RoslynUtils/SymbolExtensions.cs [hunk 1]
```diff
@@ -74,6 +74,21 @@ public static bool HasAttribute(this ImmutableArray<AttributeData> attributes, I
         return false;
     }
 
+    public static bool TryGetAttribute(this ImmutableArray<AttributeData> attributes, INamedTypeSymbol attributeType, [NotNullWhen(true)] out AttributeData? matchedAttribute)
```
Comment (JamesNK) [inline src/Shared/RoslynUtils/SymbolExtensions.cs:77]: nit: `public static bool HasAttribute(this ImmutableArray<AttributeData> attributes, INamedTypeSymbol attributeType)` implementation can be updated to call TryGetAttribute and return the bool and ignore the matched value.

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameterSource.cs [hunk 2]
```diff
@@ -14,6 +14,7 @@ internal enum EndpointParameterSource
     JsonBodyOrService,
     FormBody,
     Service,
+    KeyedService,
```
Comment (JamesNK) [inline src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameterSource.cs:17]: What happens if someone combines `Service` and `KeyedService`?

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs [hunk 3]
```diff
@@ -139,6 +139,12 @@ private void ProcessEndpointParameterSource(Endpoint endpoint, ISymbol symbol, I
         {
             Source = EndpointParameterSource.Service;
         }
+        else if (attributes.TryGetAttribute(wellKnownTypes.Get(WellKnownType.Microsoft_Extensions_DependencyInjection_FromKeyedServicesAttribute), out var keyedServicesAttribute))
+        {
+            Source = EndpointParameterSource.KeyedService;
+            var constructorArgument = keyedServicesAttribute.ConstructorArguments.FirstOrDefault();
+            AssigningCode = constructorArgument.IsNull ? string.Empty : SymbolDisplay.FormatPrimitive(constructorArgument.Value!, true, true);
```
Comment (benjaminpetit) [inline src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs:146]: `string.Empty` and `null` are not the same for keyed service:
- `null` key -> fallback to non-keyed service (case already handled by `ServiceProvider`)
- 'string.Empty' is a valid key

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameterSource.cs (hunk 2 above)
Comment (captainsafia) [reply src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameterSource.cs:17]: Good question! I can add a test with an example of the behavior here.

At the present moment, we check for the existence of the `FromService` attribute first which means that it'll always be bound as a non-keyed service.

@benjaminpetit What do you think is the correct behavior when a user annotates a parameter with both the `FromService` and `FromKeyedServices` attributes? I can see two options:

- Throw an exception stating that both attributes are not supported on a parameter
- Check only for keyed services
- Generate some code that resolves from keyed services first and then falls back to non-keyed services

I lean towards Options 1. It strikes me that the two attributes should be mutually exclusive...

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs (hunk 3 above)
Comment (captainsafia) [reply src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs:146]: Ah, good note. Let me add some test scenarios for this.

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameterSource.cs (hunk 2 above)
Comment (benjaminpetit) [reply src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameterSource.cs:17]: I also think the first option makes the most sense.

**Comment by captainsafia** [comment]
Approved via email.

**Review by halter73 — APPROVED** [review]
Should we ask for `[FromKeyedServices]` to add a `AttributeTargets.Property` target for `[AsParameters]` support? Should we define our own attribute for this purpose since `[FromServices]` does support properties and it's an unfortunate gap?

**Comment by captainsafia** [comment]
@halter73 Good question! I'd be curious to hear @benjaminpetit thoughts on the scenario as well.

IMO, the bigger issue with the lack of `AttributeUsage.Property` is the fact that this wouldn't play nice with _MVC's_ implicit surrogate model for arguments.

@benjaminpetit Is adding another AttributeUsage considered an API change here that we could do for .NET 8?

FWIW, I don't believe we'll need to do any reaction to it in ASP.NET once it's enabled since we already support recursing into the type so it's relatively cheap.

**Comment by benjaminpetit** [comment]
> @benjaminpetit Is adding another AttributeUsage considered an API change here that we could do for .NET 8?

No idea. I am not sure it's the way to go, since it would be specific for this use case (unless I am missing something).

Maybe defining your own attribute might be better, even if it might be confusing with all these attributes...

**Comment by captainsafia** [comment]
@benjaminpetit Spoke with Ben offline.

Personally, I'd prefer that we add the `AttributeUsage.Property` in the attribute in the runtime over creating a new attribute in ASP.NET Core. I want to avoid reproducing the problems we currently have with multiple variants of similar attributes throughout the stack.

For .NET 8, I'm more comfortable with an attribute that is consistent throughout all implementations but limited in functionality than inconsistent attributes with different behaviors.

**Review by benjaminpetit — APPROVED** [review]

**Comment by JamesNK** [comment]
FromKeyedService should support properties.

That would also allow DI containers that support property setters to use it.

**Pull request merged**